### PR TITLE
Consistently return StringValues.Empty instead of default for missing headers

### LINKED ISSUE #44509: (string[])Request.Headers["NonExistentHeader"] is null while (IEnumerable<string>)Request.Headers["NonExistentHeader"] is not
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Consider the following code:

```c#
string value = string.Join("\n", Request.Headers["NonExistentHeader"]);
```

It causes the following exception in runtime at method `string.Join(string separator, string[] value)`:

`ArgumentNullException: Value cannot be null. (Parameter 'value')`

For a very strange reason, the implicit cast of an empty `StringValues` struct to `string[]` type returns `null`:

```c#
string[] arr = default(StringValues);
Console.WriteLine("string[] is null: {0}", arr is null);  // prints 'string[] is null: True' message
```

While a similar cast  to IEnumerable<string> returns an expected non-null value:

```c#
var seq = (IEnumerable<string>)default(StringValues);
Console.WriteLine("IEnumerable<string> is null: {0}", seq is null);  // prints 'IEnumerable<string> is null: False' message
```

When C# compiler compiles `string.Join("\n", Request.Headers["NonExistentHeader"])` expression, it selects `string.Join(string separator, string[] value)` overload and this makes the expression to crash at runtime. While the project has nullable checks enabled, C# compiler is not able to catch that null reference error at compile time.

It looks like a bug because structs are never expected to be `null` and thus their cast operators should avoid producing `null` values as well, unless there is a very good reason for that.

The following approach allows to workaround the issue:

```c#
string value = string.Join("\n", (IEnumerable<string>)Request.Headers["NonExistentHeader"]);
```

It would be nice to have a consistent behavior here (non-null `string[]` is expected and preferred). But even if there is a `null` then C# compiler should be able to catch it at compile time.

### Expected Behavior

`(string[])default(StringValues)` aways returns a non-null value.

### Steps To Reproduce

```c#
string value = string.Join("\n", Request.Headers["NonExistentHeader"]);
```

### Exceptions (if any)

ArgumentNullException: Value cannot be null. (Parameter 'value')
at string.Join(string separator, string[] value)

### .NET Version

6.0.401

### Anything else?

Server: Kestrel

### PR DESCRIPTION
**Pull request by halter73**
Fixes #44509

### CODE REVIEW

**Comment by JamesNK** [comment]
Could there be any performance impact here? There is lots of setting, getting and resetting of headers per request.

File: src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.tt [hunk 1]
```diff
@@ -91,12 +91,25 @@ namespace Microsoft.AspNetCore.HttpSys.Internal
...
-                _<#=prop.Name#> = value;
+
+                if (value.Count > 0)
+                {
+                    _<#=prop.Name#> = value;
+                }
+                else
+                {
+                    _<#=prop.Name#> = default;
+                }
```
Comment (Tratcher) [inline src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.tt:112]:
```suggestion
                _<#=prop.Name#> = value.Count > 0 ? value : default;
```

File: src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.tt [hunk 2]
```diff
@@ -91,12 +91,25 @@ namespace Microsoft.AspNetCore.HttpSys.Internal
                     }
                     <#=MarkRead(prop.Index)#>;
                 }
-                return _<#=prop.Name#>;
+
+                if (_<#=prop.Name#>.Count > 0)
+                {
+                    return _<#=prop.Name#>;
+                }
+                return StringValues.Empty;
```
Comment (Tratcher) [inline src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.tt:99]:
```suggestion
                return _<#=prop.Name#>.Count > 0 ? _<#=prop.Name#> : StringValues.Empty;
```

**Review by Tratcher — APPROVED** [review]
As long as TryGetValue still returns false.

File: src/Shared/HttpSys/RequestProcessing/RequestHeaders.cs [hunk 3]
```diff
@@ -216,7 +221,7 @@ internal void ResetFlags()
         }
         set
         {
-            if (StringValues.IsNullOrEmpty(value))
+            if (value.Count == 0)
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/Shared/HttpSys/RequestProcessing/RequestHeaders.cs:224]: Correct. The was never the case for any of the other IHeaderDictionary implementations.

**Comment by halter73** [comment]
[HeaderCollectionBenchmark](https://github.com/dotnet/aspnetcore/blob/18f8dd7efba8a40b0c697d9317829247bb6f1251/src/Servers/Kestrel/perf/Microbenchmarks/HeaderCollectionBenchmark.cs) Microbenchmark results

### Before

#### HeaderCollectionBenchmark (Non-empty values)

|     Method |                 Type |       Mean |     Error |    StdDev |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |--------------------- |-----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
| SetHeaders | ContentLengthNumeric |  17.524 ns | 0.2337 ns | 0.2072 ns |  57,063,164.5 |     - |     - |     - |         - |
| GetHeaders | ContentLengthNumeric |   5.473 ns | 0.0458 ns | 0.0657 ns | 182,728,202.3 |     - |     - |     - |         - |
| SetHeaders |  ContentLengthString |  33.655 ns | 0.3636 ns | 0.3401 ns |  29,713,310.7 |     - |     - |     - |         - |
| GetHeaders |  ContentLengthString |  10.344 ns | 0.1054 ns | 0.0986 ns |  96,674,609.9 |     - |     - |     - |         - |
| SetHeaders |            Plaintext |  32.613 ns | 0.2117 ns | 0.1877 ns |  30,662,870.7 |     - |     - |     - |         - |
| GetHeaders |            Plaintext |   9.606 ns | 0.0880 ns | 0.0687 ns | 104,099,323.2 |     - |     - |     - |         - |
| SetHeaders |               Common | 223.445 ns | 0.1301 ns | 0.1154 ns |   4,475,382.8 |     - |     - |     - |         - |
| GetHeaders |               Common |  29.346 ns | 0.2203 ns | 0.1840 ns |  34,076,300.6 |     - |     - |     - |         - |
| SetHeaders |              Unknown | 456.815 ns | 3.8157 ns | 3.5692 ns |   2,189,068.3 |     - |     - |     - |         - |
| GetHeaders |              Unknown | 145.684 ns | 0.3838 ns | 0.3205 ns |   6,864,184.4 |     - |     - |     - |         - |

#### HeaderCollectionBenchmark.GetHeaders (Empty values)

|     Method |                 Type |       Mean |     Error |    StdDev |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |--------------------- |-----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
| GetHeaders | ContentLengthNumeric |   5.613 ns | 0.1114 ns | 0.1327 ns | 178,163,323.8 |     - |     - |     - |         - |
| GetHeaders |  ContentLengthString |   8.556 ns | 0.0838 ns | 0.0700 ns | 116,875,653.8 |     - |     - |     - |         - |
| GetHeaders |            Plaintext |  10.696 ns | 0.2026 ns | 0.1990 ns |  93,489,301.0 |     - |     - |     - |         - |
| GetHeaders |               Common |  34.490 ns | 0.1691 ns | 0.1582 ns |  28,994,028.5 |     - |     - |     - |         - |
| GetHeaders |              Unknown |  54.222 ns | 0.1896 ns | 0.1583 ns |  18,442,826.8 |     - |     - |     - |         - |


### After

#### HeaderCollectionBenchmark (Non-empty values)

|     Method |                 Type |       Mean |     Error |    StdDev |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |--------------------- |-----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
| SetHeaders | ContentLengthNumeric |  17.858 ns | 0.1414 ns | 0.1254 ns |  55,996,048.3 |     - |     - |     - |         - |
| GetHeaders | ContentLengthNumeric |   5.296 ns | 0.0842 ns | 0.0936 ns | 188,814,077.8 |     - |     - |     - |         - |
| SetHeaders |  ContentLengthString |  35.876 ns | 0.4560 ns | 0.4265 ns |  27,874,029.0 |     - |     - |     - |         - |
| GetHeaders |  ContentLengthString |  10.194 ns | 0.1119 ns | 0.1047 ns |  98,094,078.7 |     - |     - |     - |         - |
| SetHeaders |            Plaintext |  33.039 ns | 0.2072 ns | 0.1938 ns |  30,267,093.1 |     - |     - |     - |         - |
| GetHeaders |            Plaintext |   9.366 ns | 0.0912 ns | 0.0853 ns | 106,769,050.7 |     - |     - |     - |         - |
| SetHeaders |               Common | 224.466 ns | 0.2419 ns | 0.2020 ns |   4,455,008.9 |     - |     - |     - |         - |
| GetHeaders |               Common |  31.516 ns | 0.1768 ns | 0.1568 ns |  31,729,482.7 |     - |     - |     - |         - |
| SetHeaders |              Unknown | 461.242 ns | 3.2458 ns | 3.0361 ns |   2,168,059.7 |     - |     - |     - |         - |
| GetHeaders |              Unknown | 151.437 ns | 0.5270 ns | 0.4929 ns |   6,603,389.3 |     - |     - |     - |         - |

#### HeaderCollectionBenchmark.GetHeaders (Empty values)
|     Method |                 Type |       Mean |     Error |    StdDev |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |--------------------- |-----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
| GetHeaders | ContentLengthNumeric |   5.468 ns | 0.0946 ns | 0.0930 ns | 182,888,637.4 |     - |     - |     - |         - |
| GetHeaders |  ContentLengthString |   9.760 ns | 0.1130 ns | 0.1057 ns | 102,457,565.7 |     - |     - |     - |         - |
| GetHeaders |            Plaintext |  11.957 ns | 0.2128 ns | 0.1990 ns |  83,636,108.4 |     - |     - |     - |         - |
| GetHeaders |               Common |  39.305 ns | 0.1598 ns | 0.1416 ns |  25,441,734.5 |     - |     - |     - |         - |
| GetHeaders |              Unknown |  69.573 ns | 0.1160 ns | 0.0969 ns |  14,373,433.1 |     - |     - |     - |         - |

--- 

The biggest regression is in the "Unkown (Empty values) GetHeaders" microbenchmark which reads 3 known headers and then 8 unknown headers from the underlying "unknown" generic dictionary. This went from approximately 55 to 70 nanoseconds. Before this returned `default` instead of `StringValues.Empty`, so it makes sense there is some regression.

https://github.com/dotnet/aspnetcore/blob/7808fa7c1032ec64f7dda1f0d23ea6d9368e5bff/src/Servers/Kestrel/perf/Microbenchmarks/HeaderCollectionBenchmark.cs#L304-L321

The next biggest regression is in"Common (Empty values) GetHeaders" microbenchmark which reads 18 known headers from fields within the source-genned dictionary. This went from approximately 34 to 39 nanoseconds.

https://github.com/dotnet/aspnetcore/blob/7808fa7c1032ec64f7dda1f0d23ea6d9368e5bff/src/Servers/Kestrel/perf/Microbenchmarks/HeaderCollectionBenchmark.cs#L266-L290

The microbenchmark results for setting and reading non-empty values seems unnaffected. I'm plan to merge this PR considering this fixes a functional inconsistency and only causes a minor performance regression when reading a lot of unset headers. We can keep an eye on the end-to-end benchmarks after this is merged. @sebastienros

**Pull request merged**